packages/flutter/lib/src/rendering/object.dart

```diff
@@ -434,6 +434,10 @@ class PaintingContext extends ClipContext {
   /// (e.g. from opacity layer to a clip rect layer).
   /// {@endtemplate}
   ClipRectLayer? pushClipRect(bool needsCompositing, Offset offset, Rect clipRect, PaintingContextCallback painter, { Clip clipBehavior = Clip.hardEdge, ClipRectLayer? oldLayer }) {
+    if (clipBehavior == Clip.none) {
+      painter(this, offset);
```

Comment:
Why is the else clause of the needsCompositing so much more sophisticated? Perhaps you should funnel the flow to that else clause instead of interceding here (with potentially non-matching code)?

Reply:
The clip part of the code used in the needsCompositing/else doesn't need to manage the clip, but it does have a protective save/restore that I'm trying to wrap my head around. If all of the state-changing ROs protect their state change with a save/restore then that extra save/restore shouldn't be needed, right?

In which case, this immediate redirect to the painter is probably fine.

Reply:
> If all of the state-changing ROs protect their state change with a save/restore then that extra save/restore shouldn't be needed, right?

I don't think all of the ROs do, the physical model/shapes don't include a save/restore either